OS interfaces

SimonRichardson opened the pull request:
The following adds the `/servers/{server_id}/os-interface` endpoint
for listing all ports associated with a server.

Added subsequent tests to the stubs to ensure that we test the getting
the expected interfaces back from the stub server.

The is required to fulfill the need from Juju to be able to get all the
ports of a given server, to then be able to delete them when destroying
a controller.

#### Refrence

The documentation for this is here:
https://docs.openstack.org/api-ref/compute/?expanded=list-port-interfaces-detail

SimonRichardson committed:
Deprecated API Requests.

When looking over the documentation to add new API requests, some of our
existing API calls are using deprecated methods. The following changes
calls out the methods and sets them aside in a deprecated file, until we
have a plan to create a new v3 goose.

The plan would entail if we remove the old calls and provide no
alternative or if we add the new calls to allow migration forward for
others.

Either way, the first step of calling out what shouldn't be used is
probably a good path forward.

SimonRichardson committed:
Add os-interfaces to nova

The following adds the /servers/{server_id}/os-interface endpoint
for listing all ports associated with a server.

Added subsequent tests to the stubs to ensure that we test the getting
the expected interfaces back from the stub server.

The documentation for this is here:
https://docs.openstack.org/api-ref/compute/?expanded=list-port-interfaces-detail

The is required to fulfill the need from Juju to be able to get all the
ports of a given server, to then be able to delete them when destroying
a controller.

hmlanigan commented on nova/live_test.go:
```diff
@@ -747,3 +747,30 @@ func (s *LiveTests) TestCreateServerBlockDeviceMappingVolume(c *gc.C) {
+		}},
+	})
+	c.Assert(err, gc.IsNil)
+	defer s.nova.DeleteServer(entity.Id)
```
ignored returned error

SimonRichardson committed:
Expose required testing functions

Inorder to correctly test certain functionality within juju, we need
to expose this functionality to the unit tests.

hmlanigan commented on nova/nova.go:
```diff
@@ -590,3 +590,38 @@ func (c *Client) SetServerMetadata(serverId string, metadata map[string]string)
+	NetID      string        `json:"net_id,omitempty"`
+	PortID     string        `json:"port_id,omitempty"`
+	PortState  string        `json:"port_state,omitempty"`
+}
```
Just for completeness.

```suggestion
	Tag         string        `json:"tag,omitempty"`
}
```

hmlanigan commented on nova/nova.go:
```diff
@@ -590,3 +590,38 @@ func (c *Client) SetServerMetadata(serverId string, metadata map[string]string)
+// ListOSInterfaces lists all the os-interfaces (port interfaces) associated
+// with a given server.
+// https://docs.openstack.org/api-ref/compute/?expanded=list-port-interfaces-detail
+func (c *Client) ListOSInterfaces(serverId string) ([]OSInterface, error) {
```
Would it be useful to add call to encapsulate: Show Port Interface Details?  /servers/{server_id}/os-interface/{port_id}.  It provides slightly more data.

SimonRichardson replied:
I did think about doing that, I'll make a follow-up card to look at that in another PR.

hmlanigan commented on nova/nova.go:
```diff
@@ -590,3 +590,38 @@ func (c *Client) SetServerMetadata(serverId string, metadata map[string]string)
+	}
+	requestData := goosehttp.RequestData{
+		RespValue: &resp,
+	}
```
```suggestion
            ExpectedStatus: []int{http.StatusOK},
	}
```

hmlanigan reviewed (approved):
A few suggestions, but overall LGTM

SimonRichardson committed:
Expose Tag on OSInterface

Although we don't use the tag in this, we probably want to expose it for
debugging purposes.

SimonRichardson committed:
Ensure that we handle all errors

To prevent delete server error'ing out with the introduction of os
interfaces, we should verify that it doesn't return an error.
